File: go/builders_test.go
package clientattestation

import (
	"strings"
	"testing"
)

func TestAttestationExplicitTimesAndOptionalClaims(t *testing.T) {
	attester, err := Generate("ES256")
	if err != nil {
		t.Fatal(err)
	}
	instance, err := Generate("ES256")
	if err != nil {
		t.Fatal(err)
	}
	details := []map[string]any{{"type": "openid_credential"}}
	workload := map[string]any{"platform": "test"}
	attestation, err := NewClientAttestationBuilder(attester, attesterIss).
		ClientID(testClientID).
		ConfirmationJWK(instance.PublicJWK()).
		IssuedAt(1700000000).
		ExpiresAt(1700000300).
		AuthorizationDetails(details).
		Workload(workload).
		Build()
	if err != nil {
		t.Fatal(err)
	}
	_, claims := verify(t, attestation, publicKey(t, attester))
	if claims["iat"] != float64(1700000000) {
		t.Errorf("iat = %v, want 1700000000", claims["iat"])
	}
	if claims["exp"] != float64(1700000300) {
		t.Errorf("exp = %v, want 1700000300", claims["exp"])
	}
	ad, ok := claims["authorization_details"].([]any)
	if !ok || len(ad) != 1 {
		t.Errorf("authorization_details = %v", claims["authorization_details"])
	}
	wl, ok := claims["workload"].(map[string]any)
	if !ok || wl["platform"] != "test" {
		t.Errorf("workload = %v", claims["workload"])
	}
}

func TestAttestationExpiresInFromExplicitIssuedAt(t *testing.T) {
	attester, err := Generate("ES256")
	if err != nil {
		t.Fatal(err)
	}
	instance, err := Generate("ES256")
	if err != nil {
		t.Fatal(err)
	}
	attestation, err := NewClientAttestationBuilder(attester, attesterIss).
		ClientID(testClientID).
		ConfirmationKey(instance).
		IssuedAt(1700000000).
		ExpiresIn(600).
		Build()
	if err != nil {
		t.Fatal(err)
	}
	_, claims := verify(t, attestation, publicKey(t, attester))
	if claims["exp"] != float64(1700000600) {
		t.Errorf("exp = %v, want iat+600", claims["exp"])
	}
}

func TestAttestationBuilderMissingFields(t *testing.T) {
	attester, err := Generate("ES256")
	if err != nil {
		t.Fatal(err)
	}
	instance, err := Generate("ES256")
	if err != nil {
		t.Fatal(err)
	}
	cases := map[string]*ClientAttestationBuilder{
		"missing issuer": NewClientAttestationBuilder(attester, "  ").
			ClientID(testClientID).ConfirmationKey(instance).ExpiresIn(300),
		"missing client_id": NewClientAttestationBuilder(attester, attesterIss).
			ConfirmationKey(instance).ExpiresIn(300),
		"missing cnf": NewClientAttestationBuilder(attester, attesterIss).
			ClientID(testClientID).ExpiresIn(300),
		"missing expiry": NewClientAttestationBuilder(attester, attesterIss).
			ClientID(testClientID).ConfirmationKey(instance),
	}
	for name, b := range cases {
		if _, err := b.Build(); err == nil {
			t.Errorf("%s: expected an error", name)
		}
	}
}

func TestPopExplicitJtiIatAndChallenge(t *testing.T) {
	instance, err := Generate("ES256")
	if err != nil {
		t.Fatal(err)
	}
	pop, err := NewPopBuilder(instance).
		Audience(testAudience).
		Challenge("nonce-123").
		JwtID("jti-1").
		IssuedAt(1700000000).
		Build()
	if err != nil {
		t.Fatal(err)
	}
	_, claims := verify(t, pop, publicKey(t, instance))
	if claims["jti"] != "jti-1" {
		t.Errorf("jti = %v", claims["jti"])
	}
	if claims["iat"] != float64(1700000000) {
		t.Errorf("iat = %v", claims["iat"])
	}
	if claims["challenge"] != "nonce-123" {
		t.Errorf("challenge = %v", claims["challenge"])
	}
	if _, ok := claims["iss"]; ok {
		t.Error("iss must be omitted when no client_id is set")
	}
}

func TestPopRequiresAudience(t *testing.T) {
	instance, err := Generate("ES256")
	if err != nil {
		t.Fatal(err)
	}
	if _, err := NewPopBuilder(instance).Build(); err == nil {
		t.Error("expected an error for a missing audience")
	}
}

func TestDpopExplicitJtiIatNonceAndMethod(t *testing.T) {
	instance, err := Generate("ES256")
	if err != nil {
		t.Fatal(err)
	}
	dpop, err := NewDpopProofBuilder(instance).
		Method(""). // blank is ignored; htm stays POST
		Method("GET").
		URI(tokenEndpoint).
		Nonce("srv-nonce").
		JwtID("jti-2").
		IssuedAt(1700000000).
		Build()
	if err != nil {
		t.Fatal(err)
	}
	_, claims := verify(t, dpop, publicKey(t, instance))
	if claims["htm"] != "GET" {
		t.Errorf("htm = %v", claims["htm"])
	}
	if claims["jti"] != "jti-2" {
		t.Errorf("jti = %v", claims["jti"])
	}
	if claims["iat"] != float64(1700000000) {
		t.Errorf("iat = %v", claims["iat"])
	}
	if claims["nonce"] != "srv-nonce" {
		t.Errorf("nonce = %v", claims["nonce"])
	}
}

func TestDpopRequiresURI(t *testing.T) {
	instance, err := Generate("ES256")
	if err != nil {
		t.Fatal(err)
	}
	if _, err := NewDpopProofBuilder(instance).Build(); err == nil {
		t.Error("expected an error for a missing htu")
	}
}

func TestCredentialRequiresAttestationJWT(t *testing.T) {
	instance, err := Generate("ES256")
	if err != nil {
		t.Fatal(err)
	}
	if _, err := NewClientAttestationCredential("   ", instance); err == nil {
		t.Error("expected an error for a blank attestation JWT")
	}
}

func TestCredentialHeaderErrorsPropagate(t *testing.T) {
	instance, err := Generate("ES256")
	if err != nil {
		t.Fatal(err)
	}
	cred, err := NewClientAttestationCredential("a.b.c", instance)
	if err != nil {
		t.Fatal(err)
	}
	if _, err := cred.PopHeaders(testClientID, "", ""); err == nil ||
		!strings.Contains(err.Error(), "audience") {
		t.Errorf("PopHeaders without audience: err = %v", err)
	}
	if _, err := cred.DpopHeaders("POST", "", ""); err == nil ||
		!strings.Contains(err.Error(), "htu") {
		t.Errorf("DpopHeaders without uri: err = %v", err)
	}
}

File: typescript/test/builders.test.ts
import { describe, expect, it } from "vitest";
import { decodeJwt, decodeProtectedHeader } from "jose";

import {
  ClientAttestationBuilder,
  DpopProofBuilder,
  PopBuilder,
  SigningKeyPair,
} from "../src/index.js";

const ATTESTER_ISS = "https://attester.example.com";
const CLIENT_ID = "https://rp.example.com";
const AUDIENCE = "https://as.example.com";
const TOKEN_ENDPOINT = "https://as.example.com/as/token.oauth2";

describe("ClientAttestationBuilder", () => {
  it("rejects a blank issuer", async () => {
    const attester = await SigningKeyPair.generate("ES256");
    expect(() => new ClientAttestationBuilder(attester, "   ")).toThrow("issuer is required");
  });

  it("requires a client_id", async () => {
    const attester = await SigningKeyPair.generate("ES256");
    const instance = await SigningKeyPair.generate("ES256");
    await expect(
      new ClientAttestationBuilder(attester, ATTESTER_ISS)
        .confirmationKey(instance)
        .expiresIn(300)
        .build(),
    ).rejects.toThrow("client_id is required");
  });

  it("requires a confirmation key", async () => {
    const attester = await SigningKeyPair.generate("ES256");
    await expect(
      new ClientAttestationBuilder(attester, ATTESTER_ISS)
        .clientId(CLIENT_ID)
        .expiresIn(300)
        .build(),
    ).rejects.toThrow("confirmation key (cnf.jwk) is required");
  });

  it("requires an expiry", async () => {
    const attester = await SigningKeyPair.generate("ES256");
    const instance = await SigningKeyPair.generate("ES256");
    await expect(
      new ClientAttestationBuilder(attester, ATTESTER_ISS)
        .clientId(CLIENT_ID)
        .confirmationKey(instance)
        .build(),
    ).rejects.toThrow("expiry is required");
  });

  it("honours explicit issuedAt and expiresAt", async () => {
    const attester = await SigningKeyPair.generate("ES256");
    const instance = await SigningKeyPair.generate("ES256");
    const attestation = await new ClientAttestationBuilder(attester, ATTESTER_ISS)
      .clientId(CLIENT_ID)
      .confirmationKey(instance)
      .issuedAt(1_700_000_000)
      .expiresAt(1_700_000_300)
      .build();
    const payload = decodeJwt(attestation);
    expect(payload.iat).toBe(1_700_000_000);
    expect(payload.exp).toBe(1_700_000_300);
  });

  it("computes exp from issuedAt + expiresIn", async () => {
    const attester = await SigningKeyPair.generate("ES256");
    const instance = await SigningKeyPair.generate("ES256");
    const attestation = await new ClientAttestationBuilder(attester, ATTESTER_ISS)
      .clientId(CLIENT_ID)
      .confirmationKey(instance)
      .issuedAt(1_700_000_000)
      .expiresIn(600)
      .build();
    const payload = decodeJwt(attestation);
    expect(payload.exp).toBe(1_700_000_600);
  });

  it("carries authorization_details and workload when non-empty", async () => {
    const attester = await SigningKeyPair.generate("ES256");
    const instance = await SigningKeyPair.generate("ES256");
    const details = [{ type: "payment_initiation", actions: ["initiate"] }];
    const attestation = await new ClientAttestationBuilder(attester, ATTESTER_ISS)
      .clientId(CLIENT_ID)
      .confirmationKey(instance)
      .expiresIn(300)
      .authorizationDetails(details)
      .workload({ runtime: "container", region: "ap-southeast-2" })
      .build();
    const payload = decodeJwt(attestation);
    expect(payload.authorization_details).toEqual(details);
    expect(payload.workload).toEqual({ runtime: "container", region: "ap-southeast-2" });
  });

  it("omits empty authorization_details and workload", async () => {
    const attester = await SigningKeyPair.generate("ES256");
    const instance = await SigningKeyPair.generate("ES256");
    const attestation = await new ClientAttestationBuilder(attester, ATTESTER_ISS)
      .clientId(CLIENT_ID)
      .confirmationKey(instance)
      .expiresIn(300)
      .authorizationDetails([])
      .workload({})
      .build();
    const payload = decodeJwt(attestation);
    expect(payload.authorization_details).toBeUndefined();
    expect(payload.workload).toBeUndefined();
  });
});

describe("PopBuilder", () => {
  it("requires an audience", async () => {
    const instance = await SigningKeyPair.generate("ES256");
    await expect(new PopBuilder(instance).clientId(CLIENT_ID).build()).rejects.toThrow(
      "audience (aud) is required",
    );
  });

  it("carries challenge and honours jti / iat overrides", async () => {
    const instance = await SigningKeyPair.generate("ES256");
    const pop = await new PopBuilder(instance)
      .audience(AUDIENCE)
      .challenge("as-challenge")
      .jwtId("jti-1")
      .issuedAt(1_700_000_000)
      .build();
    const payload = decodeJwt(pop);
    expect(payload.challenge).toBe("as-challenge");
    expect(payload.jti).toBe("jti-1");
    expect(payload.iat).toBe(1_700_000_000);
  });

  it("omits iss and challenge when unset", async () => {
    const instance = await SigningKeyPair.generate("ES256");
    const pop = await new PopBuilder(instance).audience(AUDIENCE).challenge(null).build();
    const payload = decodeJwt(pop);
    expect(payload.iss).toBeUndefined();
    expect(payload.challenge).toBeUndefined();
  });
});

describe("DpopProofBuilder", () => {
  it("requires the uri (htu)", async () => {
    const instance = await SigningKeyPair.generate("ES256");
    await expect(new DpopProofBuilder(instance).build()).rejects.toThrow("uri (htu) is required");
  });

  it("carries nonce and honours jti / iat overrides", async () => {
    const instance = await SigningKeyPair.generate("ES256");
    const dpop = await new DpopProofBuilder(instance)
      .method("GET")
      .uri(TOKEN_ENDPOINT)
      .nonce("server-nonce")
      .jwtId("jti-2")
      .issuedAt(1_700_000_000)
      .build();
    const payload = decodeJwt(dpop);
    expect(payload.htm).toBe("GET");
    expect(payload.nonce).toBe("server-nonce");
    expect(payload.jti).toBe("jti-2");
    expect(payload.iat).toBe(1_700_000_000);
    expect(decodeProtectedHeader(dpop).typ).toBe("dpop+jwt");
  });

  it("keeps the POST default when method is null/undefined and omits an unset nonce", async () => {
    const instance = await SigningKeyPair.generate("ES256");
    const dpop = await new DpopProofBuilder(instance)
      .method(null)
      .method(undefined)
      .uri(TOKEN_ENDPOINT)
      .nonce(null)
      .build();
    const payload = decodeJwt(dpop);
    expect(payload.htm).toBe("POST");
    expect(payload.nonce).toBeUndefined();
  });
});
